simpleclient/src/main/java/io/prometheus/client/CKMSQuantiles.java

```diff
@@ -35,86 +36,158 @@
- * 
+ * <p>
+ * <p>
+ * The Approximate Quantiles Algorithm looks like this (Figure 1)
+ * <pre>
+ * Main()
+ *    foreach item v do
+ *      Insert(v);
+ *      if (Compress Condition()) then
+ *          Compress();
+ *
+ * Insert(v):
+ *    r_0:=0;
+ *    for i:=1 to s do
+ *        ri := ri−1 + gi−1;
+ *        if (v < vi) break;
+ *        add (v,1,f(ri,n)−1) to S before vi;
+ *    n++;
+ *
+ * Compress():
+ *    for i := (s−1) downto 1 do
+ *        if (gi + gi+1 + ∆i+1 ≤ f(ri, n)) then
+ *            merge ti and ti+1;
+ *
+ * Output(φ):
+ *    r0:=0;
+ *    for i := 1 to s do
+ *        ri:=ri−1+gi−1;
+ *        if(ri +gi +∆i >φn+f(φn,n)/2)
+ *            print(vi−1); break;
+ * </pre>
  */
-class CKMSQuantiles {
+public final class CKMSQuantiles { /* public for benchmark-module */
     /**
      * Total number of items in stream.
+     * Increases on every insertBatch().
      */
     private int count = 0;
 
     /**
-     * Used for tracking incremental compression.
+     * Current list of sampled items, maintained in sorted order with error
+     * bounds.
+     * <p>
+     * Note: Any algorithm that guarantees to find biased
+     * quantiles φ with error at most φεn in rank must store
+     * Ω(1 min{klog1/φ,log(εn)}) items.
+     * </p>
+     */
+    protected final LinkedList<Item> sample;
+
+    /**
+     * Used for compress condition.
+     * This is a different index than the bufferCount,
+     * because flushing is also done on `quantile.get()`,
+     * but we need to compress, regardless of reaching the bufferCount, to limit space usage.
      */
     private int compressIdx = 0;
 
     /**
-     * Current list of sampled items, maintained in sorted order with error
-     * bounds.
+     * The amount of values observed when to compress, equal to 1/2ε
      */
-    protected LinkedList<Item> sample;
+    private final int insert_threshold;
 
     /**
      * Buffers incoming items to be inserted in batch.
+     * Incoming items are buffered into blocks of size 1/2ε
      */
-    private double[] buffer = new double[500];
+    private final double[] buffer;
 
-    private int bufferCount = 0;
+    /**
+     * Tracks the current index of the buffer. Increases on insert().
+     */
+    private int bufferIdx = 0;
 
     /**
      * Array of Quantiles that we care about, along with desired error.
      */
-    private final Quantile quantiles[];
+    private final Quantile[] quantiles;
 
+    /**
+     * Set up the CKMS Quantiles. Can have 0 or more targeted quantiles defined.
+     * @param quantiles The targeted quantiles, can be empty.
+     */
     public CKMSQuantiles(Quantile[] quantiles) {
-        this.quantiles = quantiles;
+        // hard-coded epsilon of 0.1% to determine the batch size, and default epsilon in case of empty quantiles
+        double pointOnePercent = 0.001;
+        if (quantiles.length == 0) { // we need at least one for this algorithm to work
```

Comment:
Agreed, I mainly have added this because in the comments of summary it states: 
```java
  final List<Quantile> quantiles; // Can be empty, but can never be null.
```
https://github.com/prometheus/client_java/blob/master/simpleclient/src/main/java/io/prometheus/client/Summary.java#L83

I'm not sure what changes when this restriction is applied to users of Summaries that don't have specific quantiles applied to the summary.